### ISSUE TITLE
ack is slow to show first result in large directory structures

### ISSUE THREAD

**Issue by tapted** [beyondgrep/ack3#194]
In large projects (e..g https://chromium.googlesource.com/chromium/src.git/ ), it takes a while to walk the directory tree to filter the list of ~300,000 files before ack generates any output. 20 seconds is typical for this case.

In these situations, it is considerably faster to do something like
`git ls-files * | ack3 --files-from=- string`

git ls-files takes about 200ms. So ack generates output in about 3 seconds rather than 20. If, in addition, a filename regex is provided to git ls-files, e.g.
`git ls-files *.css | ack3 --css --files-from=- string`
output is basically instantaneous; <150ms.

Of course this misses any files in the directory tree that are untracked in the repository so probably isn't a good default behavior. However directory-walking in ack seems unusually slow. 
`find . -type f | ack3 --files-from=- --css string` takes 4 seconds rather than 20 when the directory structure is already "warm" in the disk cache.

It's probably difficult to restructure ack so that it can act on file list whilst it is still being built. That could help unblock the "first" output, but doesn't help the more general case.

I don't really know Perl, but maybe File::Find could be used to quickly build up the full file list before passing it to the file filtering logic.  It seems odd that File::Next::files is so much slower than File::Next::from_file though -- maybe it's something else.

**Comment by petdance** [comment]
ack doesn't filter all the files in a tree before generating output.  It will, however, have to sort all the files in a given directory before searching them if you have `--sort` turned on somewhere.  Is that the case?  You can run `ack --dump` to see what options you have set where.

**Comment by petdance** [comment]
Also, FYI, you can do `-x` instead of `--files-from=-`.  Think `-x` for "xargs".

**Comment by n1vux** [comment]
It is certainly the case that if you don't need Ack's magic file discovery
& classification, that not using it and instead  using a cached  or
explicit liste.g. git... | ack -x will be faster, and noticably so in large
source trees.

If the tree includes large irrelevant directories (binary build artifacts),
classifying all those as irrelevant takes time, and if the iterate before
the interesting subtree, worse luck for first output. Git tracks where the
interesting files are, Ack discovers at runtime.

Side note : I'll add
  git ls-files | ack -x
to Cookbook ticket #26 when I get back to desk, that's excellent sharable
idiom, thank you.

**Comment by petdance** [comment]
I made a separate ticket #195 rather than adding more to the Cookbook #26, which is getting to be bigger than I can handle.

**Comment by tapted** [comment]
> ack doesn't filter all the files in a tree before generating output. It will, however, have to sort all the files in a given directory before searching them if you have `--sort` turned on somewhere. Is that the case? You can run `ack --dump` to see what options you have set where.

I'm not using --sort[-files]. That was my first suspect for slowness, and I tried combinations of --flush and --filter without any impact to these numbers.

Note `find` is also considerably faster -- we don't need the leg-up from `git ls-files` to see a performance difference.

FTR, I gathered the numbers above with (effectively) things like the following. Note I've added "--css" to try to split out file-finding from and file-reading overheads (there are 2,947 css files).
```
time (ack --noenv --css --ignore-dir=out --ignore-dir=out_amd64-generic --flush --pager='less -R --quit-if-one-screen' webkit-focus-ring-color | head -n 1)`
ui/webui/resources/css/tabs.css:94:  outline: 5px auto -webkit-focus-ring-color;
real	0m5.692s
user	0m3.600s
sys	0m2.093s
```
vs

```
time (find -name out -prune -o -name out_amd64-generic -prune -o -name '*.css' | ack --noenv -x --flush --pager='less -R --quit-if-one-screen' webkit-focus-ring-color | head -n1)
./ui/webui/resources/css/tabs.css:94:  outline: 5px auto -webkit-focus-ring-color;
real	0m1.052s
user	0m0.565s
sys	0m0.616s
```

The results of these commands (minus the pager/head bits) are mostly identical (The `find` approach doesn't ignore ".min.css" files and prefixes paths with "./"). So I think it's safe to conclude the slowness comes from walking directories in Perl. From skimming the `ack` code I thought the file-finding and file-reading steps were distinct, but I guess that's not really the case when using iterators (sorry - I'm still a Perl n00b ;).

Using a pipe is also maybe an unfair comparison since that allows the file-finding and file-reading to be parallelised in two processes. But (thinking simplistically) that should only improve the performance 2x, not 7x.

The `find` approach is also considerably faster when gathering all results, not just the first:
```
 (master) tapted@~/devel/git/chromium/src: time (ack --noenv --css --ignore-dir=out --ignore-dir=out_amd64-generic --flush webkit-focus-ring-color > res2.txt)
real	0m19.261s
user	0m12.166s
sys	0m7.090s
 (master) tapted@~/devel/git/chromium/src: time (find . -name out -prune -o -name out_amd64-generic -prune -o -name '*.css' -print | ack --noenv -x --css --flush webkit-focus-ring-color > res1.txt)
real	0m2.644s
user	0m1.397s
sys	0m1.555s
```

**Comment by petdance** [comment]
Tomorrow I will make a test that you can run that I hope can narrow it down. I think there is something on your machine that is different, maybe in Perl. I get instant results when I try it on the chromium repo. 

I’ll look tomorrow.  Thanks for your help on this.

**Comment by petdance** [comment]
What's an ack command you've tried that takes 20 seconds?  The 20 seconds to get results is what concerns me most right now.

`find` is always going to be faster than ack, without a doubt, but 20 seconds to get results doesn't match what I'm seeing at all.

**Comment by n1vux** [comment]
For a worst-case inode layout with a very large tree which is sparse in "interesting" files, I can easily believe the simplest invocation of Ack _could be_ very very slow in returning the first answer. It's a matter of luck in inode arrangement in the directories, which is usually non-reproducible, but i can imagine that `git clone` might even make it reproducible!
There are enough GIT projects that I expect at least one of them to be structured so its clones are worst-case inode layout for our find.

The exemplar project's [.gitignore](https://chromium.googlesource.com/chromium/src.git/+/refs/heads/master/.gitignore) file provides some suggestion for how large a tree Ack is walking fruitlessly. There's potentially a _lot_ of uninteresting artifacts in there. So this could easily be the one I expect above.

If the `inode/readdir(2)` order in the tree is such that Ack's  `File::Next` wanders down the ignorable subtrees depth first in a large forest of `.gitignore` dirs, directories full of files that either we reject as not searchable or worse, we accept as searchable but they're boring and have no real chance of having matches because they're e.g. log wallpaper of CI builds or whatever, not real source code but still fine scannable ascii, well yes it could be a very very very long time to first light.

Basic hints for using *`ack`* on a _truly large_ heterogeneous filesystem tree are
 - be as explicit as possible about what file types you expect/want matches in (--cc vs --rust, --perl, --yaml ... or make your own type sets as needed)
 - be explicit about what sub-dir(s) the targets are in so we don't read `./logs` as well as `./src` ... and possibly `./logs` first if `git clone` gave that the earlier inode; good news is that like (modern) find, you can name several starter dirs.
 - put anything you can in `.ackrc` and/or commandline `--ignore-dir` and `--ignore-file` (similar to `.gitignore` but lets you ignore things git might care about and vice versa) 
 - use `ack -x` aka `ack --files-from -`  to use a pre-computed static list or an easy to compute list; the OP's `git ls-files|` is excellent example of this, but even more fun is `git diff --stat  origin | cut -d' ' -f2 | ack -x ...` to only look at the changed files.  I'll also do a `find ... > files.lst ;`  and use that precomputed list for several `ack -x`'s  with different patterns on the same set of files.

(If those hints aren't in Docs or Docs issues already, they should be ...)

Yes *Find* will always be faster. 
Doubly so: 
It's written in C (vs Perl) and classifies items only as Directory or File, it doesn't try to classify Binary or Source, let alone Perl vs CC vs Rust vs CSV. 
Ack is not a better Find; although sometimes 'ack -g' will do the trick with its more fulsome RE, it isn't a total substitute. (And it really doesn't want to match .gif names, since it couldn't search into them.)

(And that's before the advantage that Find is on a buffered pipe, not running as a co-routine.  But  if it's walking down large blind alleys depth first, parallel find will not expedite first result; if block-buffered, it will even delay it slightly.)

(As noted, sorted would ruin everything in big trees, whether piped or co-routined finding; if you do `find | sort | xargs`, you destroy buffered advantage there too. Sort is the bottleneck of pipes.) 


I am unclear if we need to make any changes except Doc changes for this. 

In this case, limiting the scope of the search is far more useful than anything we could possibly do generally, and probably in any other case where the filetree is large enough and pessimal enough to make ack noticeably slow to first answer, it will be likewise. 

OTOH, this is bad from a least-surprise point of view, so needs to be well documented edge case where Ack needs more hints than one might expect.
